[SPARK-20619][ML] StringIndexer supports multiple ways to order label

actuaryzhang opened the pull request:
## What changes were proposed in this pull request?

StringIndexer maps labels to numbers according to the descending order of label frequency. Other types of ordering (e.g., alphabetical) may be needed in feature ETL.  For example, the ordering will affect the result in one-hot encoding and RFormula. 

This PR proposes to support other ordering methods and we add a parameter `stringOrderType` that supports the following four options:
- 'frequencyDesc': descending order by label frequency (most frequent label assigned 0)
- 'frequencyAsc': ascending order by label frequency (least frequent label assigned 0)
- 'alphabetDesc': descending alphabetical order
- 'alphabetAsc': ascending alphabetical order

The default is still descending order of label frequency, so there should be no impact to existing programs.

## How was this patch tested?
new test

Commit:
StringIndexer supports multiple ways of label ordering

SparkQA commented:
**[Test build #76517 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76517/testReport)** for PR 17879 at commit [`ffd0cfc`](https://github.com/apache/spark/commit/ffd0cfc755586402c0f22e4458149eed16f98010).

actuaryzhang commented:
@jkbradley @MLnick @holdenk @pnpritchard @yanboliang @sethah @imatiach-msft @srowen 

An example to illustrate the idea:
```
val data = Seq((0, "b"), (1, "b"), (2, "c"), (3, "a"), (4, "a"), (5, "b"))
val df = data.toDF("id", "label")
val indexer = new StringIndexer()
      .setInputCol("label")
      .setOutputCol("labelIndex")
df.show
+---+-----+
| id|label|
+---+-----+
|  0|    b|
|  1|    b|
|  2|    c|
|  3|    a|
|  4|    a|
|  5|    b|
+---+-----+
```

Below is the result corresponding to the different types of label ordering. 
```
indexer.setStringOrderType("freq_desc").fit(df).transform(df)
+---+-----+----------+
| id|label|labelIndex|
+---+-----+----------+
|  0|    b|       0.0|
|  1|    b|       0.0|
|  2|    c|       2.0|
|  3|    a|       1.0|
|  4|    a|       1.0|
|  5|    b|       0.0|
+---+-----+----------+

indexer.setStringOrderType("freq_asc").fit(df).transform(df)
+---+-----+----------+
| id|label|labelIndex|
+---+-----+----------+
|  0|    b|       2.0|
|  1|    b|       2.0|
|  2|    c|       0.0|
|  3|    a|       1.0|
|  4|    a|       1.0|
|  5|    b|       2.0|
+---+-----+----------+

indexer.setStringOrderType("alphabet_desc").fit(df).transform(df)
+---+-----+----------+
| id|label|labelIndex|
+---+-----+----------+
|  0|    b|       1.0|
|  1|    b|       1.0|
|  2|    c|       0.0|
|  3|    a|       2.0|
|  4|    a|       2.0|
|  5|    b|       1.0|
+---+-----+----------+

indexer.setStringOrderType("alphabet_asc").fit(df).transform(df)
+---+-----+----------+
| id|label|labelIndex|
+---+-----+----------+
|  0|    b|       1.0|
|  1|    b|       1.0|
|  2|    c|       2.0|
|  3|    a|       0.0|
|  4|    a|       0.0|
|  5|    b|       1.0|
+---+-----+----------+

```

holdenk commented:
What would be some common cases where alphabet ordering would be needed?

actuaryzhang commented:
@holdenk The main motivation for this PR is that the behavior of StringIndexer will affect OneHotEncoder, RFormula and models estimated based on these transformers. There have been a few desired improvement in RFormula that could not be done without the change in StringIndexer.

One use case for alphabetical ordering is to make comparison of Spark model results to that in R, which drops the first alphabetical value in one-hot encoding. Right now, even though we do lots of comparisons between Spark and R, we lack comparisons involving String features because the encoding is different. There is already a [JIRA|https://issues.apache.org/jira/browse/SPARK-14659]. 

Another motivation for this PR is to support ascending order by label frequency. This is also related to one-hot encoding. In practical applications of regression type models, it is almost always better to set the most frequent label as the reference level (i.e., drop the most frequent label in OneHotEncoding) for better interpretability. Right now, the behavior is the opposite and has made it very difficult to interpret results. 

I think  the flexibility of different ordering will benefit a lot the downstream feature transformers and model estimators. Does this make sense?

SparkQA commented:
**[Test build #76517 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76517/testReport)** for PR 17879 at commit [`ffd0cfc`](https://github.com/apache/spark/commit/ffd0cfc755586402c0f22e4458149eed16f98010).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76517/
Test PASSed.

actuaryzhang commented:
@yanboliang Since you have reported a few issues due to different encoding between Spark and R (e.g., #SPARK-14659 and #SPARK-14657), probably you could add some comments?

viirya commented on mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala:
```diff
@@ -59,6 +59,28 @@ private[feature] trait StringIndexerBase extends Params with HasInputCol with Ha
+  final val stringOrderType: Param[String] = new Param(this, "stringOrderType",
+    "The method used to order values of input column. " +
+      s"Supported options: ${StringIndexer.supportedStringOrderType.mkString(", ")}.",
+    (value: String) => StringIndexer.supportedStringOrderType.contains(value.toLowerCase))
```
Use `ParamValidators.inArray`?

actuaryzhang replied:
@viirya `ParamValidators.inArray` does not allow case-insensitive validation, does it?

viirya replied:
Yeah, I originally thought you'd change to case-sensitive. It looks good to me.

viirya commented on mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala:
```diff
@@ -107,11 +135,15 @@ class StringIndexer @Since("1.4.0") (
+      .select(col($(inputCol)).cast(StringType))
+      .rdd.map(_.getString(0))
+    val labels = $(stringOrderType) match {
+      case "freq_desc" => values.countByValue().toSeq.sortBy(-_._2).map(_._1).toArray
```
Seems this setting is case-insensitive because your param validator doesn't care it, `$(stringOrderType)` might be upper-case here. We may make it case-sensitive or do `toLowerCase` here?

actuaryzhang replied:
@viirya Great catch. Thanks!

Commit:
address review comments and fix style

actuaryzhang commented:
@viirya Thanks much for your comments. Made a new commit to address them.

viirya commented on mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala:
```diff
@@ -59,6 +59,28 @@ private[feature] trait StringIndexerBase extends Params with HasInputCol with Ha
+
+  /** @group getParam */
+  @Since("2.2.0")
+  def getStringOrderType: String = $(stringOrderType)
```
I checked other ML classes. Looks like for a case-insensitive setting, we may do toLowerCase in its public API:

    def getStringOrderType: String = $(stringOrderType).toLowerCase

And you can use `getStringOrderType` below instead of `$(stringOrderType).toLowerCase` in `fit`.

actuaryzhang replied:
@viirya Which ML classes were you referring to? I was told not to change the raw values in the getters in other PRs  #16675.

SparkQA commented:
**[Test build #76521 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76521/testReport)** for PR 17879 at commit [`97e020f`](https://github.com/apache/spark/commit/97e020f4aba1afcf45c1b10de09d6fbba1551918).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76521/
Test PASSed.

actuaryzhang commented:
ping @yanboliang @felixcheung 
This is needed for one-hot encoding to be consistent with R, therefore enabling directly comparison of Spark results to R. Could you guys please take a look? Thanks.

felixcheung commented on mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala:
```diff
@@ -131,6 +163,12 @@ object StringIndexer extends DefaultParamsReadable[StringIndexer] {
   private[feature] val KEEP_INVALID: String = "keep"
   private[feature] val supportedHandleInvalids: Array[String] =
     Array(SKIP_INVALID, ERROR_INVALID, KEEP_INVALID)
+  private[feature] val FREQ_DESC: String = "freq_desc"
```
is there any prior standard for these names like `freq_desc`?

actuaryzhang replied:
@felixcheung I did not find any prior standard, and am open to suggestion for better names. 
Maybe better use `frequency_desc` or `count_desc`?

felixcheung replied:
@gatorsmile thought?

felixcheung reviewed (commented):
LGTM - I want to double check https://github.com/apache/spark/pull/17879/files#r115116845
And given where we are now we should change this from 2.2 to 2.3

actuaryzhang commented:
@felixcheung Thanks. I will update the annotation.

viirya commented:
@actuaryzhang There seems something wrong with Github's webpage, so I can't directly reply the above comment. `ALSModelParams.getColdStartStrategy` is one example.

Commit:
address comments- spell out freq and update annotation and toLowerCase

SparkQA commented:
**[Test build #76619 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76619/testReport)** for PR 17879 at commit [`ba34043`](https://github.com/apache/spark/commit/ba340437fee99f848dfa88eab2e10d87651eab0a).

SparkQA commented:
**[Test build #76619 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76619/testReport)** for PR 17879 at commit [`ba34043`](https://github.com/apache/spark/commit/ba340437fee99f848dfa88eab2e10d87651eab0a).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76619/
Test FAILed.

Commit:
fix style

actuaryzhang commented:
Thanks much @felixcheung and @viirya. I have addressed your comments. 
- update from 2.2 to 2.3
- change `freq_desc` to `frequency_desc`. 
- move toLowerCase to the getter method. 
Please let me know if there is anything needed. Thanks!

SparkQA commented:
**[Test build #76621 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76621/testReport)** for PR 17879 at commit [`ff9b1d6`](https://github.com/apache/spark/commit/ff9b1d66873eb8cad1a4a13f323555da2706a849).

Commit:
fix annotation

SparkQA commented:
**[Test build #76624 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76624/testReport)** for PR 17879 at commit [`07198d9`](https://github.com/apache/spark/commit/07198d9bb45a54d3c257ad37e772cc31154ffcb6).

gatorsmile commented on mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala:
```diff
@@ -131,6 +167,12 @@ object StringIndexer extends DefaultParamsReadable[StringIndexer] {
+  private[feature] val FREQ_DESC: String = "frequency_desc"
+  private[feature] val FREQ_ASC: String = "frequency_asc"
+  private[feature] val ALPHABET_DESC: String = "alphabet_desc"
+  private[feature] val ALPHABET_ASC: String = "alphabet_asc"
```
Normally, we do not use underscore in the names. `lowerCamelCase` is our rules for naming.

actuaryzhang replied:
@gatorsmile Thanks much for the suggestion. Changed them to lowerCamelCase. 
@felixcheung Any additional suggestions?

Commit:
use camel case

Commit:
remove extra import

SparkQA commented:
**[Test build #76628 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76628/testReport)** for PR 17879 at commit [`53381ea`](https://github.com/apache/spark/commit/53381ea6ba41cc26ed89a6fc42252f7126198d9f).

SparkQA commented:
**[Test build #76621 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76621/testReport)** for PR 17879 at commit [`ff9b1d6`](https://github.com/apache/spark/commit/ff9b1d66873eb8cad1a4a13f323555da2706a849).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76621/
Test PASSed.

SparkQA commented:
**[Test build #76624 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76624/testReport)** for PR 17879 at commit [`07198d9`](https://github.com/apache/spark/commit/07198d9bb45a54d3c257ad37e772cc31154ffcb6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

viirya commented:
LGTM

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76624/
Test PASSed.

felixcheung commented on mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala:
```diff
@@ -59,6 +59,29 @@ private[feature] trait StringIndexerBase extends Params with HasInputCol with Ha
+    "how to order labels of string column. " +
+    "The first label after ordering is assigned an index of 0. " +
+    s"Supported options: ${StringIndexer.supportedStringOrderType.mkString(", ")}.",
+    ParamValidators.inArray(StringIndexer.supportedStringOrderType))
```
so we are going to case sensitive then?

actuaryzhang replied:
@felixcheung  Right. It does not quite make sense to be case-sensitive now given that we now use camel case.

SparkQA commented:
**[Test build #76628 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76628/testReport)** for PR 17879 at commit [`53381ea`](https://github.com/apache/spark/commit/53381ea6ba41cc26ed89a6fc42252f7126198d9f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76628/
Test PASSed.

actuaryzhang commented:
@felixcheung @gatorsmile @MLnick @jkbradley @holdenk @yanboliang @srowen @sethah 
Would you please take another look and let me know any additional suggestions? Thanks much!

felixcheung reviewed (approved):
LGTM, did another pass.
@yanboliang @jkbradley do you want to have a look?

actuaryzhang commented:
@felixcheung Thanks much for your review. 
@yanboliang @jkbradley Since there are two approvals, could you guys take a look and merge if it's good? We really need this for a couple of SparkR related issues, e.g., SPARK-14659 and SPARK-14657. Thanks much!

actuaryzhang commented:
@felixcheung 
It would be great if you could help merge this. I could address comments (if any) in a future PR.
This seems a pretty straightforward change that removes a big blocker. I will continue working on the RFormula side once this merges in and fix the SparkR issues related to string ordering. 

Thanks much!

felixcheung commented:
merged to master.

actuaryzhang commented:
Thanks much!
